cpp/include/cudf/table/row_operators.cuh

```diff
@@ -222,13 +223,75 @@ class element_equality_comparator {
   bool nulls_are_equal;
 };
 
+namespace detail {
+
+/**
+ * @brief Helper functor to check if two types `LhsType` and `RhsType` are comparable for equality.
+ */
+template <typename LhsType>
+struct equality_comparable_functor {
+  template <typename RhsType>
+  bool __device__ operator()() const
+  {
+    return cudf::is_equality_comparable<LhsType, RhsType>();
+  }
+};
+
+/**
+ * @brief Helper functor to check if a specified column `rhs_col` can be compared for equality
+ * with type `LhsType`.
+ */
+struct lhs_equality_comparable_functor {
+  template <typename LhsType>
+  bool __device__ operator()(column_device_view rhs_col) const
+  {
+    return cudf::type_dispatcher(rhs_col.type(), equality_comparable_functor<LhsType>{});
+  }
+};
+
+/**
+ * @brief Checks if two columns have types that permit equality comparisons
+ *
+ * @param lhs Left column for equality comparisons
+ * @param rhs Right column for equality comparisons
+ * @return true If `lhs` and `rhs` columns have types whose values may be compared for equality
+ * @return false Otherwise.
+ */
+bool __device__ columns_are_equality_comparable(column_device_view lhs, column_device_view rhs)
+{
+  return cudf::type_dispatcher(lhs.type(), lhs_equality_comparable_functor{}, rhs);
+}
+
+/**
+ * @brief Checks if corresponding columns in `lhs` and `rhs` tables may be compared for equality
+ *
+ * @param lhs Left table, whose columns are to be compared for equality.
+ * @param rhs Right table, whose columns are to be compared for equality.
+ * @return true If *all* columns in `lhs` permit equality comparisons with corresponding columns in
+ * `rhs`.
+ * @return false Otherwise.
+ */
+bool all_columns_are_equality_comparable(table_device_view lhs, table_device_view rhs)
+{
+  auto columns_begin = thrust::make_counting_iterator<size_type>(0);
+  auto columns_end   = columns_begin + lhs.num_columns();
+  auto eq_comparable = [lhs, rhs] __device__(auto col_idx) {
+    return columns_are_equality_comparable(lhs.column(col_idx), rhs.column(col_idx));
+  };
+  return thrust::all_of(thrust::seq, columns_begin, columns_end, eq_comparable);
```

Comment:
So the lambda is a device lambda, but this is running in host code. Furthermore, we shouldn't be launching a kernel to check this. That's expensive.

Reply:
`lhs.column(col_idx)` is a `__device__` function. 

The only recourse I can think of is to change `row_equality_comparator` to operate on `table_view`. That looks deeply intrusive. There are a lot of call sites that seem to wrap tightly around the `table_device_view` bit.

Reply:
Agreed. Lets defer the `row_equality_comparator` changes, but keep the `equality_comparable` check and just put it at a higher level in groupby for now.